ignition-server/controllers/tokensecret_controller.go

```diff
@@ -0,0 +1,147 @@
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "updateEvent"), e.ObjectNew, TokenSecretAnnotation)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "createEvent"), e.Object, TokenSecretAnnotation)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "deleteEvent"), e.Object, TokenSecretAnnotation)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return processIfMatchesAnnotation(ctrl.LoggerFrom(ctx).WithValues("predicate", "genericEvent"), e.Object, TokenSecretAnnotation)
+		},
+	}
+}
+func (r *TokenSecretReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	log := ctrl.Log.WithName("Setting up token controller")
+	log.Info("SetupWithManager", "ns", os.Getenv("MY_NAMESPACE"))
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Secret{}).WithEventFilter(tokenSecretAnnotationPredicate(ctx)).
+		Complete(r)
+}
+
+func processIfMatchesAnnotation(logger logr.Logger, obj client.Object, annotation string) bool {
+	kind := strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind)
+	log := logger.WithValues("namespace", obj.GetNamespace(), kind, obj.GetName())
+
+	if _, ok := obj.GetAnnotations()[annotation]; ok {
+		log.Info("Resource matches annotation, will attempt to map resource")
+		return true
+	}
+
+	log.Info("Resource does not match annotation, will not attempt to map resource")
+	return false
+}
+
+func (r *TokenSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling")
+
+	tokenSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, req.NamespacedName, tokenSecret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found", "request", req.String())
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "error token Secret")
+		return ctrl.Result{}, err
+	}
+
+	if !tokenSecret.DeletionTimestamp.IsZero() {
+		// When this reconciler watch the deletion event
+		// and tries to get the Resource it might already be gone
+		// therefore this is unlikely to be reached.
+		// This is just a best effort to synchronous cleanup.
+		// The PayloadStore expiring mechanism takes care of consistently delete expired entries.
+		PayloadStore.Delete(string(tokenSecret.Data[TokenSecretTokenKey]))
+		return ctrl.Result{}, nil
+	}
+
+	token := string(tokenSecret.Data[TokenSecretTokenKey])
+	if _, ok := PayloadStore.Get(token); ok {
+		log.Info("Payload found in cache")
+		return ctrl.Result{}, nil
+	}
+
+	releaseImage := string(tokenSecret.Data[TokenSecretReleaseKey])
+	payload, err := r.IgnitionProvider.GetPayload(ctx, releaseImage)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error getting ignition payload: %v", err)
+	}
+	if payload == nil {
```

Comment:
IgnitionProvider.GetPayload is synchronous and all-or-nothing, so not sure this check is needed